### PR TITLE
GEODE-7309: uplift lucene to 7.1.0

### PR DESCRIPTION
**Pull request by mkevo**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-lucene/build.gradle [hunk 1]
```diff
@@ -71,6 +71,7 @@ dependencies {
 
   upgradeTestImplementation(project(':geode-dunit'))
   upgradeTestImplementation('commons-io:commons-io')
+  upgradeTestImplementation('org.apache.lucene:lucene-backward-codecs')
```
Comment (onichols-pivotal) [inline geode-lucene/build.gradle:74]: just checking, is backward-codecs really only required by tests, not end-users too?  I would naively assume it needs to be shipped with Geode as well.

**Review by onichols-pivotal — APPROVED** [review]

File: geode-lucene/build.gradle (hunk 1 above)
Comment (mkevo) [reply geode-lucene/build.gradle:74]: I read some articles about Lucene and its similarities and only find that this backward-codecs is needed for test. So I'm not sure taht we need to ship it with Geode.

File: geode-lucene/src/upgradeTest/java/org/apache/geode/cache/lucene/LuceneSearchWithRollingUpgradeTestBase.java [hunk 2]
```diff
@@ -295,22 +295,26 @@ void verifyLuceneQueryResults(String regionName, int expectedRegionSize)
...
-      } catch (Exception ex) {
-        if (!ex.getCause().getMessage().contains("currently indexing")) {
-          throw ex;
-        }
-        if (--retryCount == 0) {
-          throw ex;
-        }
-      }
-    }
+    await().untilAsserted(() -> assertThat(isIndexingFinished(luceneQuery)).isTrue());
```
[the remark this answers is not in the document]
Comment (kirklund) [reply geode-lucene/src/upgradeTest/java/org/apache/geode/cache/lucene/LuceneSearchWithRollingUpgradeTestBase.java:298]: You can do some experimenting to see the failure messages with something like this:
```
  @Test
  public void foo() {
    await().atMost(2, TimeUnit.SECONDS).ignoreExceptions().untilAsserted(() -> {
      throw new IllegalStateException("foo");
    });
  }
```

File: geode-lucene/src/upgradeTest/java/org/apache/geode/cache/lucene/LuceneQueryWithDifferentVersions.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      locator1.invoke(
+          invokeStartLocator(hostName, locatorPorts[0], getLocatorPropertiesPre91(locatorString)));
+      locator2.invoke(
+          invokeStartLocator(hostName, locatorPorts[1], getLocatorPropertiesPre91(locatorString)));
+      invokeRunnableInVMs(invokeCreateCache(getSystemProperties(locatorPorts)), server1, server2);
+
+      // Locators before 1.4 handled configuration asynchronously.
+      // We must wait for configuration configuration to be ready, or confirm that it is disabled.
+      locator1.invoke(
+          () -> await()
```
Comment (kirklund) [inline geode-lucene/src/upgradeTest/java/org/apache/geode/cache/lucene/LuceneQueryWithDifferentVersions.java:65]: I would get rid of the `!InternalLocator.getLocator().getConfig().getEnableClusterConfiguration()` unless you're parameterizing it to run both with and without cluster config. Then you can simplify it to:
```
await().untilAsserted(() -> {
  assertThat(InternalLocator.getLocator().isSharedConfigurationRunning()).isTrue();
}
```

File: geode-lucene/src/upgradeTest/java/org/apache/geode/cache/lucene/RollingUpgradeQueryReturnsCorrectResultsAfterClientAndServersAreRolledOverAllBucketsCreated.java [hunk 4]
```diff
@@ -31,6 +32,7 @@
 public class RollingUpgradeQueryReturnsCorrectResultsAfterClientAndServersAreRolledOverAllBucketsCreated
     extends LuceneSearchWithRollingUpgradeDUnit {
 
+  @Ignore
```
Comment (kirklund) [inline geode-lucene/src/upgradeTest/java/org/apache/geode/cache/lucene/RollingUpgradeQueryReturnsCorrectResultsAfterClientAndServersAreRolledOverAllBucketsCreated.java:35]: Don't commit this with `@Ignore` on a test.

**Review by kirklund — CHANGES_REQUESTED** [review]

File: geode-docs/tools_modules/lucene_integration.html.md.erb [hunk 5]
```diff
@@ -334,6 +334,7 @@ Lucene indexes created for all members.
 # <a id="LuceneRandC" class="no-quick-link"></a>Requirements and Caveats
 
 - Join queries between regions are not supported.
+- Executing queries is possible after all members are at the same version.
```
Comment (karensmolermiller) [inline geode-docs/tools_modules/lucene_integration.html.md.erb:337]: This statement is not quite specific enough. I'm guessing that all members must be at the same minor version number (not necessarily the same patch version within the minor).  I'm also guessing that this is just for Lucene queries.

If I'm correct in my guesses, an attempt at better wording might be
> All cluster members must be running the same minor version in order to execute Lucene queries.

**Review by karensmolermiller — CHANGES_REQUESTED** [review]

File: geode-docs/tools_modules/lucene_integration.html.md.erb (hunk 5 above)
Comment (mkevo) [reply geode-docs/tools_modules/lucene_integration.html.md.erb:337]: Yes, you are right! Thanks!

File: geode-lucene/src/upgradeTest/java/org/apache/geode/cache/lucene/LuceneQueryWithDifferentVersions.java (hunk 3 above)
Comment (mkevo) [reply geode-lucene/src/upgradeTest/java/org/apache/geode/cache/lucene/LuceneQueryWithDifferentVersions.java:65]: Thanks!

**Review by rhoughton-pivot — APPROVED** [review]

File: geode-core/src/main/java/org/apache/geode/internal/cache/InternalCacheForClientAccess.java [hunk 6]
```diff
@@ -1225,6 +1226,14 @@ public void unlockDiskStore(String diskStoreName) {
 
   }
 
+  @Override
+  public boolean hasMemberOlderThan(KnownVersion version) {
+    return getMembers().stream()
```
[the remark this answers is not in the document]
Comment (mkevo) [reply geode-core/src/main/java/org/apache/geode/internal/cache/InternalCacheForClientAccess.java:1231]: Thanks for the suggestion! 
As both `GemFireCacheImpl` and `InternalCacheForClientAccess` implements `InternalCache`, I think that we need to add method body in both classes.

File: geode-lucene/src/upgradeTest/java/org/apache/geode/cache/lucene/RollingUpgradeQueryReturnsCorrectResultsAfterClientAndServersAreRolledOverAllBucketsCreated.java (hunk 4 above)
Comment (mkevo) [reply geode-lucene/src/upgradeTest/java/org/apache/geode/cache/lucene/RollingUpgradeQueryReturnsCorrectResultsAfterClientAndServersAreRolledOverAllBucketsCreated.java:35]: Upgrading to the Lucene 7.1.0 this test is not supported as formatting is changed between versions. In that case we add note to the documentation so all cluster members must be on the same version in order to execute Lucene queries.

**Review by karensmolermiller — APPROVED** [review]

**Comment by karensmolermiller** [comment]
Docs change looks good to me now.

**Comment by mkevo** [comment]
The new PR is created, rebased on the latest develop branch, and removed **_1** from the branch name.
 #6351

**Pull request closed without merging**